### PR TITLE
add apcomp package

### PR DESCRIPTION
**Pull request by cyrush**

### CODE REVIEW

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+
+class Apcomp(Package):
+    """A multi use-case image compositor"""
+
+    homepage = 'https://github.com/Alpine-DAV/ap_compositor'
+    git      = 'https://github.com/Alpine-DAV/ap_compositor.git'
+    url      = "https://github.com/Alpine-DAV/ap_compositor/releases/download/v0.0.1/apcomp-v0.0.1.tar.gz"
+
+    maintainers = ['mclarsen','cyrush']
```
Comment (ax3l) [inline var/spack/repos/builtin/packages/apcomp/package.py]: PEP8 probably needs this:
```suggestion
    maintainers = ['mclarsen', 'cyrush']
```

**Comment by cyrush** [comment]
fake8 ==> flake8 :-)

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+        #######################
+        # Compiler Info
+        #######################
+        c_compiler = env["SPACK_CC"]
+        cpp_compiler = env["SPACK_CXX"]
+
+        #######################################################################
+        # By directly fetching the names of the actual compilers we appear
+        # to doing something evil here, but this is necessary to create a
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/apcomp/package.py]: (minor) I don't think this is evil (in other words, FWIW while it is good to mention why you are fetching the true compiler names, you don't have to phrase it as an admission).

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    git      = 'https://github.com/Alpine-DAV/ap_compositor.git'
+    url      = "https://github.com/Alpine-DAV/ap_compositor/releases/download/v0.0.1/apcomp-v0.0.1.tar.gz"
+
+    maintainers = ['mclarsen', 'cyrush']
+
+    version('master', branch='master', submodules='True')
+    version('0.0.1', sha256="cbf85fe58d5d5bc2f468d081386cc8b79861046b3bb7e966edfa3f8e95b998b2", preferred=True)
+
+    variant('openmp', default=True, description='Build with openmp support')
+    variant('mpi', default=True, description='Build with openmp support')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/apcomp/package.py]: This and the `openmp` variant have the same description.

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    depends_on("mpi", when="+mpi")
+
+    root_cmakelists_dir = 'src'
+
+    def install(self, spec, prefix):
+        """
+        Build and install APComp
+        """
+        with working_dir('spack-build', create=True):
+            py_site_pkgs_dir = None
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/apcomp/package.py]: Is there a reason you initialize this to `None` to pass it to `create_host_config` vs. just passing `None` to `create_host_config` directly?

Also, while `create_host_config` takes a `py_site_pkgs_dir` parameter, it doesn't do anything with that argument. Is this preparing for some future PR?

**Review by scheibelp — COMMENTED** [review]
I noticed a few minor things - would you have a chance to update this?

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    """A multi use-case image compositor"""
+
+    homepage = 'https://github.com/Alpine-DAV/ap_compositor'
+    git      = 'https://github.com/Alpine-DAV/ap_compositor.git'
+    url      = "https://github.com/Alpine-DAV/ap_compositor/releases/download/v0.0.1/apcomp-v0.0.1.tar.gz"
+
+    maintainers = ['mclarsen', 'cyrush']
+
+    version('master', branch='master', submodules='True')
+    version('0.0.1', sha256="cbf85fe58d5d5bc2f468d081386cc8b79861046b3bb7e966edfa3f8e95b998b2", preferred=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/apcomp/package.py]:
```suggestion
    version('0.0.1', sha256="cbf85fe58d5d5bc2f468d081386cc8b79861046b3bb7e966edfa3f8e95b998b2")
```
Spack already defaults to the newest stable release

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            print("Configuring APComp...")
+            cmake(*cmake_args)
+            print("Building APComp...")
+            make()
+            print("Installing APComp...")
+            make("install")
+            # install copy of host config for provenance
+            install(host_cfg_fname, prefix)
+
+    def create_host_config(self, spec, prefix, py_site_pkgs_dir=None):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/apcomp/package.py]: Is there any reason we can't extend `CMakePackage` instead of adding all this custom logic?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/apcomp/package.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    """
+    if vtype is None:
+        if value == "ON" or value == "OFF":
+            vtype = "BOOL"
+        else:
+            vtype = "PATH"
+    return 'set({0} "{1}" CACHE {2} "")\n\n'.format(name, value, vtype)
+
+
+class Apcomp(Package):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/apcomp/package.py:26]: Why not call the package `ap-compositor`?

File: var/spack/repos/builtin/packages/apcomp/package.py (hunk 4 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/apcomp/package.py]: this is a hold over from using another package as a starting point, we can prune.

File: var/spack/repos/builtin/packages/apcomp/package.py (hunk 3 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/apcomp/package.py]: thanks, I'll fix this

File: var/spack/repos/builtin/packages/apcomp/package.py (hunk 6 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/apcomp/package.py]: we need to capture the relevant cmake config values for provenance, debugging, and to allow folks to build when developing against an existing set of spack-built tpls.

Another perspective: The magic folded into the CMakePackage via inheritance shifts over time -- it's hard to keep up with choices that being made as things evolve and are convolved together without extensive vetting.

File: var/spack/repos/builtin/packages/apcomp/package.py (hunk 2 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/apcomp/package.py]: fixing this as well, I am glad it is no longer viewed as evil :-)

**Comment by cyrush** [comment]
Thanks for the suggestions, I pushed updates to address. 
For the name, I would prefer to keep the name `apcomp`.

**Review by adamjstewart — APPROVED** [review]

**Comment by cyrush** [comment]
thanks!

**Pull request merged**